Remove nl sys argument defaulting

lindsayad opened the pull request:
Closes #24137

lindsayad committed:
Remove nl sys argument defaulting

Among other things:
- Change UserObject::_nl_sys to UserObject::_sys

Closes #24137

lindsayad commented:
Keep open plz

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1740560/) on 2ae6e20 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24373/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1743576/) on 78c7417 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.66%</td>
      <td>85.65%</td>
      <td>-0.01%</td>
      <td>91.69%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90468</td>
      <td>90516</td>
      <td>+48</td>
      <td>331</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15151</td>
      <td>15169</td>
      <td>+18</td>
      <td>30</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.86%</td>
      <td>89.89%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3830</td>
      <td>3839</td>
      <td>+9</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>432</td>
      <td>432</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/contact)

### External petsc solver

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.45%</td>
      <td>85.45%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>370</td>
      <td>370</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>63</td>
      <td>63</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/external_petsc_solver)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/external_petsc_solver)

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.46%</td>
      <td>88.46%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3679</td>
      <td>3679</td>
      <td>-</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>480</td>
      <td>480</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/heat_conduction)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.96%</td>
      <td>83.96%</td>
      <td>+0.00%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10981</td>
      <td>10987</td>
      <td>+6</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2098</td>
      <td>2099</td>
      <td>+1</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/navier_stokes)

### Peridynamics

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>78.11%</td>
      <td>78.11%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3011</td>
      <td>3011</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>844</td>
      <td>844</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/peridynamics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/peridynamics)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.86%</td>
      <td>85.86%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12041</td>
      <td>12041</td>
      <td>-</td>
      <td>14</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1983</td>
      <td>1983</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/phase_field)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.69%</td>
      <td>95.69%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10779</td>
      <td>10779</td>
      <td>-</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>485</td>
      <td>485</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/porous_flow)

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.36%</td>
      <td>93.36%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4061</td>
      <td>4063</td>
      <td>+2</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>289</td>
      <td>289</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/ray_tracing)

### Richards

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>94.59%</td>
      <td>94.59%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3252</td>
      <td>3252</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>186</td>
      <td>186</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/richards)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/richards)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.19%</td>
      <td>91.19%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7631</td>
      <td>7631</td>
      <td>-</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>737</td>
      <td>737</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/stochastic_tools)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.67%</td>
      <td>84.67%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>25606</td>
      <td>25606</td>
      <td>-</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4635</td>
      <td>4635</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/tensor_mechanics)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.66%</td>
      <td>90.66%</td>
      <td>-</td>
      <td>75.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13281</td>
      <td>13281</td>
      <td>-</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1368</td>
      <td>1368</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/thermal_hydraulics)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>507928</th>
      <th colspan=4>#24373 78c741</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.11%</td>
      <td>82.11%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6677</td>
      <td>6676</td>
      <td>-1</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1455</td>
      <td>1455</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24373/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24373/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 66.67% is less than the suggested 90.0%
- `thermal_hydraulics` new line coverage rate 75.00% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1744877/) on f0b7478 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Public app tests](https://civet.inl.gov/job/1817031/) on 35fd3e5 : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1817030/) on 35fd3e5 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation public apps](https://civet.inl.gov/job/1817029/) on 35fd3e5 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/1817028/) on 35fd3e5 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation controlled apps](https://civet.inl.gov/job/1817027/) on 35fd3e5 : invalidated by @lindsayad

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1817025/) on 35fd3e5 : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1828359/) on b08e205 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/1828357/) on b08e205 : invalidated by @lindsayad

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1828370/) on b08e205 : invalidated by @lindsayad

lindsayad committed:
Override nlConverged instead of converged in framework problems

Subchannel (and possibly others) override nlConverged so we have
to make sure their override does something

lindsayad commented:
@grmnptr as the multi-system man, I think you'd be a good person to look at this

lindsayad commented:
I count 11 requested CCB reviewers on this PR 😄  If I could get a review from one of you at some point, I would appreciate it

grmnptr commented:
I'm on it!

grmnptr commented on modules/navier_stokes/include/base/NSFVBase.h:
```diff
@@ -934,6 +934,9 @@ NSFVBase<BaseType>::validParams()
                               "von_karman_const_0 mixing_length_delta turbulent_prandtl",
                               "Turbulence");
 
+  params.addParam<NonlinearSystemName>(
```
pretty cool!

grmnptr commented on framework/src/actions/CopyNodalVarsAction.C:
```diff
@@ -52,15 +52,14 @@ CopyNodalVarsAction::act()
       // Is this a NonlinearSystem variable or an AuxiliarySystem variable?
       if (_current_task == "copy_nodal_vars")
-        system = &_problem->getNonlinearSystemBase();
+        system = &_problem->getNonlinearSystemBase(/*nl_sys=*/0);
```
Also, for example xfem thm and some other calls default for fetching things from nonlinear system 0. So maybe a warning could be nice for those routines.

lindsayad replied:
I don't really want to take on the responsibility of adding warnings for everything that doesn't work with multisystem. I would have to go back through every translation unit I've modified and add that which I don't have the appetite for

grmnptr reviewed (changes requested):
Pertty cool stuff! A few questions, comments below.

lindsayad commented on framework/include/outputs/ConsoleUtils.h:
```diff
@@ -57,7 +57,7 @@ std::string outputAuxiliarySystemInformation(FEProblemBase & problem);
  * Output the Nonlinear system information
  */
-std::string outputNonlinearSystemInformation(FEProblemBase & problem);
+std::string outputNonlinearSystemInformation(FEProblemBase & problem, unsigned int nl_sys_num);
```
Pedantically someone looking at this API to determine whether they want to call it or not will not care whether the `unsigned int` argument is `const` or not because it is passed-by-value ... the caller does not have to worry about whether the method modifies the incoming argument because it is passed-by-value. And then in the implementation, if you actually don't modify the argument, you can mark the argument as `const` if you want since it does not alter the signature of the method. So if you look at the `.C` implementation you will likely see that I've done that

lindsayad replied:
But I do see some value in conveying the implementation in the header

lindsayad commented on framework/include/problems/FEProblem.h:
```diff
@@ -29,7 +29,7 @@ class FEProblem : public FEProblemBase
   virtual void setInputParametersFEProblem(InputParameters & parameters) override;
 
-  NonlinearSystem & getNonlinearSystem(unsigned int nl_sys_num = 0) override;
+  virtual NonlinearSystem & getNonlinearSystem(unsigned int nl_sys_num) override;
```
Yes but I think it's easier to navigate the doxygen when the `virtual` is on the front

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -170,17 +170,17 @@ class FEProblemBase : public SubProblem, public Restartable
 
   // DEPRECATED METHOD
-  void setCouplingMatrix(CouplingMatrix * cm, unsigned int nl_sys = 0);
+  void setCouplingMatrix(CouplingMatrix * cm, unsigned int nl_sys);
```
I would prefer to. I would prefer it if another PR can remove it if it's no longer being used

grmnptr replied:
Sounds good to me!

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2032,13 +2032,9 @@ class FEProblemBase : public SubProblem, public Restartable
-   * @return the nonlinear system number corresponding to the provided \p nl_sys_name
-   */
-  unsigned int nlSysNum(const NonlinearSystemName & nl_sys_name) const;
+  virtual std::size_t numNonlinearSystems() const override { return _num_nl_sys; }
```
It does but it helps me to more quickly parse what functions are virtual when perusing doxygen. I've missed noting a function as virtual in the past when it was only labeled as override

grmnptr replied:
Okay, makes sense

lindsayad committed:
Address review

- `const unsigned int nl_sys_num` in class definition
- fix doxygen string: `i` -> `nl_sys_num`

lindsayad committed:
THREAD_ID -> const THREAD_ID in the problems

lindsayad committed:
More const tid and nl_sys_num param args

lindsayad committed:
Update code for non-defaulted eigen methods

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1849353/) on 31360c2 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation controlled apps](https://civet.inl.gov/job/1849355/) on 31360c2 : invalidated by @lindsayad

lindsayad commented:
agggh apps keep adding new code that uses the default arguments. I've updated Griffin twice now and will need to do a third round...

dschwen commented:
> agggh apps keep adding new code that uses the default arguments. I've updated Griffin twice now and will need to do a third round...

Just break it

lindsayad commented on framework/include/problems/SubProblem.h:
```diff
@@ -90,15 +90,20 @@ class SubProblem : public Problem
    * @return whether the given \p nl_sys_num is converged
    */
-  virtual bool nlConverged(unsigned int nl_sys_num) = 0;
+  virtual bool nlConverged(unsigned int) { return converged(/*nl_sys_num=*/0); }
```
fixed

lindsayad committed:
Don't use friends to allow actions access to the nonlinear system

lindsayad commented on framework/include/predictors/Predictor.h:
```diff
@@ -68,4 +68,6 @@ class Predictor : public MooseObject, public Restartable
   /// Timestep tolerance from Transient executioner
   const Real & _timestep_tolerance;
+
+  friend class SetupPredictorAction;
```
this design allowed restricting access to the `_nl` member to just the actions. But it's probably not a big deal to allow public access. Changed to this in the most recent commit

dschwen replied:
Yeah, I don't understand the `friend` hate here. It seems misplaced as the "fix" seems clearly worse.

GiudGiud replied:
I quite like the "SetupXXXAction" being friends with the XXX base class personally

dschwen replied:
Even better would be the pass key idiom! :rofl:

GiudGiud replied:
the added job security from no one being able to read your code is so worth it

grmnptr replied:
Okay, I see the majority of people (who know C++ much better than I do) are for the friend-based implementation here so I am officially withdrawing my objection. Also, thanks for the detailed explanation!

dschwen commented on framework/include/materials/DerivativeMaterialInterface.h:
```diff
@@ -469,7 +469,7 @@ DerivativeMaterialInterface<T>::validateCoupling(const MaterialPropertyName & ba
   // iterate over all variables in the both the non-linear and auxiliary system (optional)
   validateCouplingHelper<U, is_ad>(
-      prop_name, c, _dmi_fe_problem.getNonlinearSystemBase().system(), missing);
+      prop_name, c, _dmi_fe_problem.getNonlinearSystemBase(/*nl_sys=*/0).system(), missing);
```
Is tehre a way to get the proper system from `T` (which would be one of `Material`, `Kernel`, `AuxKernel` etc.)

grmnptr replied:
The only way I can think of right now is through these parameters of a MooseObject:
```
  params.addPrivateParam<SystemBase *>("_sys", nullptr);
  params.addPrivateParam<SystemBase *>("_nl_sys", nullptr);
```
I am not sure if these are set for every object type. But if it isn't you can default to 0, I suppose.

lindsayad replied:
> Is there a way to get the proper system from T (which would be one of Material, Kernel, AuxKernel etc.)?

Often-times there is, however, with materials there is not. Materials are system agnostic, e.g. they can compute quantities from variables in all systems

lindsayad committed:
Revert "Don't use friends to allow actions access to the nonlinear system"

This reverts commit 5342e71df063abfa13416fd778fdd4f873728852.

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1852323/) on 3fd146d : invalidated by @lindsayad

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/1852321/) on 3fd146d : invalidated by @lindsayad

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1852334/) on 3fd146d : invalidated by @lindsayad

moosebuild commented:
Job [Conda (Rocky)](https://civet.inl.gov/job/1852335/) on 3fd146d : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1852344/) on 3fd146d wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 82.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 34.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 64.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1852334/) on 3fd146d : invalidated by @lindsayad

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1852325/) on 3fd146d : invalidated by @lindsayad

moosebuild commented:
Job [Documentation controlled apps](https://civet.inl.gov/job/1852327/) on 3fd146d : invalidated by @lindsayad

lindsayad commented:
coverage failure due to base with no coverage report

lindsayad commented:
thanks @grmnptr for your review!
